File: app/src/lib/websocket.ts
/**
 * WebSocket client for real-time updates
 */

import { io, Socket } from 'socket.io-client';

const WS_BASE_URL = import.meta.env.VITE_WS_URL || 'ws://localhost:8000/api/v1/ws';

class WebSocketManager {
  private sockets: Map<string, Socket> = new Map();
  private listeners: Map<string, Set<(data: any) => void>> = new Map();

  connect(channel: string, token: string): Socket {
    if (this.sockets.has(channel)) {
      return this.sockets.get(channel)!;
    }

    const socket = io(`${WS_BASE_URL}/${channel}`, {
      auth: { token },
      transports: ['websocket'],
      reconnection: true,
      reconnectionDelay: 1000,
      reconnectionAttempts: 5,
    });

    socket.on('connect', () => {
      console.log(`Connected to ${channel} WebSocket`);
    });

    socket.on('disconnect', () => {
      console.log(`Disconnected from ${channel} WebSocket`);
    });

    socket.on('error', (error) => {
      console.error(`WebSocket error on ${channel}:`, error);
    });

    // Handle incoming messages
    socket.on('message', (data) => {
      this.notifyListeners(channel, data);
    });

    this.sockets.set(channel, socket);
    return socket;
  }

  disconnect(channel: string): void {
    const socket = this.sockets.get(channel);
    if (socket) {
      socket.disconnect();
      this.sockets.delete(channel);
    }
  }

  disconnectAll(): void {
    this.sockets.forEach((socket) => socket.disconnect());
    this.sockets.clear();
  }

  subscribe(channel: string, callback: (data: any) => void): () => void {
    if (!this.listeners.has(channel)) {
      this.listeners.set(channel, new Set());
    }
    this.listeners.get(channel)!.add(callback);

    // Return unsubscribe function
    return () => {
      this.listeners.get(channel)?.delete(callback);
    };
  }

  private notifyListeners(channel: string, data: any): void {
    const listeners = this.listeners.get(channel);
    if (listeners) {
      listeners.forEach((callback) => callback(data));
    }
  }

  // Specific channel helpers
  connectAlerts(token: string): Socket {
    return this.connect('alerts', token);
  }

  connectDetections(token: string, cameraId?: string): Socket {
    const channel = cameraId ? `detections?camera_id=${cameraId}` : 'detections';
    return this.connect(channel, token);
  }

  connectCameraStream(cameraId: string, token: string): Socket {
    return this.connect(`cameras/${cameraId}/stream`, token);
  }

  connectSystem(token: string): Socket {
    return this.connect('system', token);
  }
}

export const wsManager = new WebSocketManager();
export default wsManager;

File: surveillance-system/backend/app/api/v1/endpoints/websocket.py
"""WebSocket endpoints for real-time updates."""

import asyncio
import json
from typing import Set, Dict
from uuid import UUID

from fastapi import APIRouter, WebSocket, WebSocketDisconnect, Depends, Query
from fastapi.security import HTTPBearer

from app.core.redis import get_redis, publish, subscribe
from app.core.config import settings
from app.api.deps import get_current_user

router = APIRouter()
security = HTTPBearer()

# Store active WebSocket connections
class ConnectionManager:
    """Manage WebSocket connections."""
    
    def __init__(self):
        self.active_connections: Dict[str, Set[WebSocket]] = {
            "alerts": set(),
            "detections": set(),
            "cameras": set(),
            "system": set()
        }
    
    async def connect(self, websocket: WebSocket, channel: str):
        """Accept and store connection."""
        await websocket.accept()
        if channel in self.active_connections:
            self.active_connections[channel].add(websocket)
    
    def disconnect(self, websocket: WebSocket, channel: str):
        """Remove connection."""
        if channel in self.active_connections:
            self.active_connections[channel].discard(websocket)
    
    async def broadcast(self, message: dict, channel: str):
        """Broadcast message to all connections in channel."""
        if channel not in self.active_connections:
            return
        
        disconnected = set()
        for connection in self.active_connections[channel]:
            try:
                await connection.send_json(message)
            except Exception:
                disconnected.add(connection)
        
        # Clean up disconnected clients
        for conn in disconnected:
            self.active_connections[channel].discard(conn)


manager = ConnectionManager()


@router.websocket("/alerts")
async def alerts_websocket(
    websocket: WebSocket,
    token: str = Query(...)
):
    """WebSocket for real-time alert notifications."""
    # TODO: Validate token
    await manager.connect(websocket, "alerts")
    
    try:
        # Subscribe to Redis channel
        redis = get_redis()
        pubsub = redis.pubsub()
        await pubsub.subscribe("alerts:updates")
        
        while True:
            # Check for Redis messages
            message = await pubsub.get_message(ignore_subscribe_messages=True, timeout=1)
            if message:
                data = json.loads(message["data"])
                await websocket.send_json(data)
            
            # Check for client ping
            try:
                data = await asyncio.wait_for(
                    websocket.receive_text(),
                    timeout=0.1
                )
                if data == "ping":
                    await websocket.send_text("pong")
            except asyncio.TimeoutError:
                pass
            
    except WebSocketDisconnect:
        manager.disconnect(websocket, "alerts")
        await pubsub.unsubscribe("alerts:updates")


@router.websocket("/detections")
async def detections_websocket(
    websocket: WebSocket,
    camera_id: UUID = Query(None),
    token: str = Query(...)
):
    """WebSocket for real-time detection updates."""
    await manager.connect(websocket, "detections")
    
    try:
        redis = get_redis()
        pubsub = redis.pubsub()
        await pubsub.subscribe("detections:updates")
        
        while True:
            message = await pubsub.get_message(ignore_subscribe_messages=True, timeout=1)
            if message:
                data = json.loads(message["data"])
                
                # Filter by camera if specified
                if camera_id and data.get("camera_id") != str(camera_id):
                    continue
                
                await websocket.send_json(data)
            
            try:
                data = await asyncio.wait_for(
                    websocket.receive_text(),
                    timeout=0.1
                )
                if data == "ping":
                    await websocket.send_text("pong")
            except asyncio.TimeoutError:
                pass
            
    except WebSocketDisconnect:
        manager.disconnect(websocket, "detections")
        await pubsub.unsubscribe("detections:updates")


@router.websocket("/cameras/{camera_id}/stream")
async def camera_stream_websocket(
    websocket: WebSocket,
    camera_id: UUID,
    token: str = Query(...)
):
    """WebSocket for camera stream proxy (HLS/WebRTC signaling)."""
    await manager.connect(websocket, "cameras")
    
    try:
        while True:
            # Receive signaling messages
            data = await websocket.receive_text()
            message = json.loads(data)
            
            # Handle different message types
            if message.get("type") == "offer":
                # TODO: Forward to WebRTC gateway
                pass
            elif message.get("type") == "ice-candidate":
                # TODO: Forward ICE candidate
                pass
            
            # Echo back for now
            await websocket.send_json({
                "type": "ack",
                "camera_id": str(camera_id)
            })
            
    except WebSocketDisconnect:
        manager.disconnect(websocket, "cameras")


@router.websocket("/system")
async def system_websocket(
    websocket: WebSocket,
    token: str = Query(...)
):
    """WebSocket for system health and status updates."""
    await manager.connect(websocket, "system")
    
    try:
        while True:
            # Send periodic health updates
            await websocket.send_json({
                "type": "health",
                "timestamp": datetime.utcnow().isoformat(),
                "status": "healthy"
            })
            
            # Wait for next update interval
            await asyncio.sleep(settings.WS_HEARTBEAT_INTERVAL)
            
    except WebSocketDisconnect:
        manager.disconnect(websocket, "system")


# Helper function to broadcast messages
async def broadcast_alert(alert_data: dict):
    """Broadcast alert to all connected clients."""
    await manager.broadcast({
        "type": "alert",
        "data": alert_data
    }, "alerts")


async def broadcast_detection(detection_data: dict):
    """Broadcast detection to all connected clients."""
    await manager.broadcast({
        "type": "detection",
        "data": detection_data
    }, "detections")
